### PR TITLE
Retry on IPv6 listing

### PR DESCRIPTION
**Pull request by CallMeFoxie**
## Description
A fix for https://github.com/projectcalico/calico/issues/2553

Detailed info what is happening is in the link, this is probably the easiest fix of all. I've been testing the code (or at least its calico 3.3 branch) for a while now and it seems to be working great (actually triggered several retries). 

```
Apr 24 09:10:17 <hostname redacted> kubelet[13796]: 2019-04-24 09:10:17.326 [INFO][19173] network.go 212: No IPv6 set on interface, retrying.. ContainerID="f5276c4c91358bfa76ee56c654e5ce83e3aca22dccec5104f0c581632a87510a" Namespace="default" Pod="test-ipv6-fvirtbase-1556089800-lggbv" WorkloadEndpoint="<hostname redacted>-k8s-test--ipv6--fvirtbase--1556089800--lggbv-eth0"
```

The bug was there always, just a recent change either in kubelet or something has made the logic faster and triggered the bug :).

IMHO it should be backported to previous branches as well as not everybody runs the latest Calico (we have 3.3 in production atm). 

Thanks for looking at this,
Ashley

## Release Note

```release-note
None required
```

### CODE REVIEW

File: internal/pkg/utils/network_linux.go [hunk 1]
```diff
@@ -187,20 +187,33 @@ func DoNetworking(
...
 			}
 
-			// No need to add a dummy next hop route as the host veth device will already have an IPv6
-			// link local address that can be used as a next hop.
-			// Just fetch the address of the host end of the veth and use it as the next hop.
-			addresses, err := netlink.AddrList(hostVeth, netlink.FAMILY_V6)
-			if err != nil {
-				logger.Errorf("Error listing IPv6 addresses for the host side of the veth pair: %s", err)
-				return err
+			var err error
```
Comment (fasaxc) [inline internal/pkg/utils/network_linux.go]: Please can you add a comment explaining why the retry is needed?   Something like `// Sometimes the IPv6 address takes a short time to appear after the sysctls are enabled.  Retry a few times.`

Comment (CallMeFoxie) [reply internal/pkg/utils/network_linux.go]: added :)

File: internal/pkg/utils/network_linux.go [hunk 2]
```diff
@@ -187,20 +187,35 @@ func DoNetworking(
...
+					logger.Errorf("Error listing IPv6 addresses for the host side of the veth pair: %s", err)
+				}
+
+				if len(addresses) < 1 {
+					// If the hostVeth doesn't have an IPv6 address then this host probably doesn't
+					// support IPv6. Since a IPv6 address has been allocated that can't be used,
+					// return an error.
+					err = fmt.Errorf("failed to get IPv6 addresses for host side of the veth pair")
+				}
+				if err == nil {
```
Comment (aojea) [inline internal/pkg/utils/network_linux.go]: what err is checking this condition?

Comment (CallMeFoxie) [reply internal/pkg/utils/network_linux.go]: Whoops misread.
This is the one checking if we can break after less than 10 retries.

Comment (fasaxc) [reply internal/pkg/utils/network_linux.go]: It terminates the retry loop if there is no error at both 198 and 207

Comment (CallMeFoxie) [reply internal/pkg/utils/network_linux.go]: Yeah, check my 2nd reply, I was looking at the wrong err.

It just makes it not retry 10 times (as in that case there may have just been time.Sleep(500ms)) but continue as fast as possible.

Comment (CallMeFoxie) [reply internal/pkg/utils/network_linux.go]: Should I still change it to flip the `if len(addresses)` or is this ok? :) (missed that due to the edit)

Comment (fasaxc) [reply internal/pkg/utils/network_linux.go]: I think it's OK, it's how the old code worked so no _need_ to change it.

**Comment by caseydavenport** [comment]
^ I've submitted some backport PRs to v3.7, v3.6, and v3.5.

**Comment by CallMeFoxie** [comment]
Thanks, I will either upgrade our ancient 3.3 to something newer or keep the update I tested it with :)

Can I change the release note somehow? Or will it get auto-set from the cherry picks?

**Comment by caseydavenport** [comment]
I think the release note is all set - it should get pulled in from the cherry picks.

**Pull request merged**